### PR TITLE
SLD: export options png

### PR DESCRIPTION
**Pull request by elpaso**
Followup recent SLD improvements https://github.com/qgis/QGIS/pull/51277 this PR adds a few small fixes and the new option to export polygon point pattern and line pattern fills as PNG ExternalGraphic images in the SLD.

By choosing the new SLD export option (only available for polygon layers), the symbol layer export implementation will try to convert the symbol to a PNG tile that can be used as an external graphic symbol by the SLD polygon fill symbolizer.

The PNG image tiles are saved in the same directory of the exported SLD.

sld_png_export

## Examples

Generated tile and rendering on Deegree

SLD_Single_Polygon_hatches

immagine

SLD_Single_Polygon_point_pattern_315

immagine

### CODE REVIEW

**Comment by nyalldawson** [comment]
Is there any reason not to make this option always on?

**Comment by elpaso** [comment]
> Is there any reason not to make this option always on?

Yes: one reason is that most of the SLD export code is currently optimized to use GeoServer SLD vendor extensions, the plan in the long run is to allow the user to choose if she wants the SLD export to be compatible with plain vanilla SLD or to enable GeoServer (or other vendor) extensions.

The other reason is backward compatibility: by not checking the option we fall back to the current implementation.

I think that in case the target server does support vector drawing with vendor extensions it is generally better to use it instead of a rasterized background tile.

**Comment by nyalldawson** [comment]
> Yes: one reason is that most of the SLD export code is currently optimized to use GeoServer SLD vendor extensions, the plan in the long run is to allow the user to choose if she wants the SLD export to be compatible with plain vanilla SLD or to enable GeoServer (or other vendor) extensions.

Sounds reasonable. How about revising the GUI to make the checkbox "Enable GeoServer extensions" then? The tooltip could say that it results in a higher quality conversion through the use of the GeoServer vendor extensions. 

To me that's an immediately understandable option for users, vs the current checkbox which requires the user to read and understand the lengthy tooltip.

File: src/core/symbology/qgssymbollayerutils.h [hunk 1]
```diff
@@ -897,6 +897,14 @@ class CORE_EXPORT QgsSymbolLayerUtils
      */
     static QgsStringMap evaluatePropertiesMap( const QMap<QString, QgsProperty> &propertiesMap, const QgsExpressionContext &context );
 
+    /**
+     * Calculate the minimum size in pixels of a symbol tile given the symbol \a width and \a height and the grid rotation \a angle in radians.
+     * The method makes approximations and can modify \a angle in order to generate the smallest possible tile.
+     * \note Angle must be >= 0 and < 2 * PI
+     * \since QGIS 3.30
```
Comment (nyalldawson) [inline src/core/symbology/qgssymbollayerutils.h:904]: If you add \returns and \param arguments here, then sipify will correctly handle the INOUT argument in the docstring.

File: python/core/auto_generated/symbology/qgssymbollayer.sip.in [hunk 2]
```diff
@@ -1313,6 +1313,19 @@ The ``rings`` argument optionally specifies a list of polygon rings to render as
     void setAngle( double angle );
     double angle() const;
 
+    virtual QImage toTiledPatternImage( ) const;
+%Docstring
+Renders the symbol layer to an image that can be used as a seamless pattern fill
```
Comment (nyalldawson) [inline python/core/auto_generated/symbology/qgssymbollayer.sip.in:1318]:
```suggestion
Renders the symbol layer as an image that can be used as a seamless pattern fill
```

File: python/core/auto_generated/symbology/qgssymbollayerutils.sip.in [hunk 3]
```diff
@@ -997,6 +997,18 @@ Evaluates a map of properties using the given ``context`` and returns a variant
 .. versionadded:: 3.18
 %End
 
+    static QSize tileSize( int width, int height, double &angleRad /In,Out/ );
+%Docstring
+Calculate the minimum size in pixels of a symbol tile given the symbol ``width`` and ``height`` and the grid rotation ``angle`` in radians.
```
Comment (nyalldawson) [inline python/core/auto_generated/symbology/qgssymbollayerutils.sip.in:1002]: angle is clockwise or counter clockwise?

File: src/core/qgsmaplayer.h [hunk 4]
```diff
@@ -1174,7 +1174,7 @@ class CORE_EXPORT QgsMapLayer : public QObject
      * \see loadSldStyle()
      * \since QGIS 3.30
      */
-    virtual QString saveSldStyleV2( const QString &uri, bool &resultFlag, const QgsSldExportContext &exportContext ) const;
+    virtual QString saveSldStyleV2( bool &resultFlag, const QgsSldExportContext &exportContext ) const;
```
Comment (nyalldawson) [inline src/core/qgsmaplayer.h:1177]:
```suggestion
    virtual QString saveSldStyleV2( bool &resultFlag SIP_OUT, const QgsSldExportContext &exportContext ) const;
```
?

File: src/core/symbology/qgssymbollayerutils.cpp [hunk 5]
```diff
@@ -5096,3 +5097,234 @@ QgsStringMap QgsSymbolLayerUtils::evaluatePropertiesMap( const QMap<QString, Qgs
   }
   return properties;
 }
+
+QSize QgsSymbolLayerUtils::tileSize( int width, int height, double &angleRad )
+{
+
+  // Precondition
+  Q_ASSERT( angleRad >= 0 && angleRad < M_PI * 2 );
```
Comment (nyalldawson) [inline src/core/symbology/qgssymbollayerutils.cpp:5105]: This is not so nice for pyqgis use -- couldn't we just mod the value automatically to ensure this condition is met?

File: src/core/symbology/qgssymbollayerutils.h [hunk 6]
```diff
@@ -897,6 +897,14 @@ class CORE_EXPORT QgsSymbolLayerUtils
      */
     static QgsStringMap evaluatePropertiesMap( const QMap<QString, QgsProperty> &propertiesMap, const QgsExpressionContext &context );
 
+    /**
+     * Calculate the minimum size in pixels of a symbol tile given the symbol \a width and \a height and the grid rotation \a angle in radians.
```
Comment (nyalldawson) [inline src/core/symbology/qgssymbollayerutils.h:901]: What do you mean by "grid rotation" here? I'm concerned that this logic is ONLY used for sld export, and is a logical duplication of the code in QgsLinePatternFillSymbolLayer::applyPattern ... ultimately giving us two very complex "line pattern to image" code paths, with their own sets of unique bugs.

File: src/core/qgsfileutils.cpp [hunk 7]
```diff
@@ -529,3 +529,23 @@ QStringList QgsFileUtils::splitPathToComponents( const QString &input )
...
+{
+  if ( ! QFileInfo::exists( path ) )
+  {
+    return path;
+  }
+
+  QFileInfo info { path };
+  const QString suffix { info.completeSuffix() };
+  const QString pathPattern { QString( suffix.length() > 0 ? path.chopped( suffix.length() + 1 ) : path ).append( QStringLiteral( "_%1." ) ).append( suffix ) };
+  int i { 1 };
```
[the remark this answers is not in the document]
Comment (elpaso) [reply src/core/qgsfileutils.cpp:543]: Yeah, maybe clearer.

File: src/core/symbology/qgssymbollayerutils.h (hunk 6 above)
Comment (elpaso) [reply src/core/symbology/qgssymbollayerutils.h:901]: Grid rotation means the rotation of the whole symbol layer (I'll change the docs to clarify), opposed to the individual marker rotation. It makes sense for point pattern fills and it is currently only used by point pattern fills, line pattern fills are simpler to calculate and do not use this method but a faster and simple formula.

File: src/core/symbology/qgssymbollayerutils.cpp (hunk 5 above)
Comment (elpaso) [reply src/core/symbology/qgssymbollayerutils.cpp:5105]: Ok, I'll move the logic from client code into this method

File: src/core/qgsfileutils.h [hunk 8]
```diff
@@ -224,6 +224,19 @@ class CORE_EXPORT QgsFileUtils
...
+     * Creates a unique file path name from a desired path by appending "_<n>" (where "<n>" is an integer number) before the file suffix.
+     *
+     * E.g. if "/path/my_image.png" already exists "/path/my_image_1.png" (and "_2", "_3" etc.) will be checked until a file path that does not already exist is found.
+     *
+     * \param path the desired path.
+     * \return the unmodified path if path is already unique or the new path with "_<n>" (where "<n>" is an integer number) appended to the file name before the suffix.
+     * \note This function does not make any check on path validity and write permissions.
+     * \since QGIS 3.30
+     */
+    static QString uniquePath( const QString &path );
```
[the remark this answers is not in the document]
Comment (elpaso) [reply src/core/qgsfileutils.h:238]: done

File: src/core/qgsfileutils.cpp [hunk 9]
```diff
@@ -529,3 +529,23 @@ QStringList QgsFileUtils::splitPathToComponents( const QString &input )
...
+QString QgsFileUtils::uniquePath( const QString &path )
+{
+  if ( ! QFileInfo::exists( path ) )
+  {
+    return path;
+  }
+
+  QFileInfo info { path };
+  const QString suffix { info.completeSuffix() };
+  const QString pathPattern { QString( suffix.length() > 0 ? path.chopped( suffix.length() + 1 ) : path ).append( QStringLiteral( "_%1." ) ).append( suffix ) };
```
[the remark this answers is not in the document]
Comment (elpaso) [reply src/core/qgsfileutils.cpp]: done

**Comment by elpaso** [comment]
> > Yes: one reason is that most of the SLD export code is currently optimized to use GeoServer SLD vendor extensions, the plan in the long run is to allow the user to choose if she wants the SLD export to be compatible with plain vanilla SLD or to enable GeoServer (or other vendor) extensions.
> 
> Sounds reasonable. How about revising the GUI to make the checkbox "Enable GeoServer extensions" then? The tooltip could say that it results in a higher quality conversion through the use of the GeoServer vendor extensions.
> 
> To me that's an immediately understandable option for users, vs the current checkbox which requires the user to read and understand the lengthy tooltip.

The logic to actually enable/disable vendor extensions is not yet implemented, the SLD export code assumes GeoServer in a lot of different places  and it is not trivial to check them all.

My intention here is to lay the API foundation to be able in the future to configure the SLD export depending on the target.

It is not correct that by enabling PNG the SLD process will skip all GeoServer extensions, some of them may still be enabled when exporting other symbols.

The final GUI in the future will probably be something like this:

- Combo box to select the vendor extensions with [No extensions, GeoServer extensions, XYZ Server extensions ...]
- Combo  box to select polygon fill export tiles mode [Avoid tiles when possible, PNG Tiles, SVG Tiles]

Note that the my (possibly wrong) rationale behind the choice to not make tiles export by default is twofold: first I would like to keep the old behavior the default (because I guess that GeoServer is the most common target), second having the PNG as sidecar files makes it more difficult to use the SLD on the target because you don't just copy-paste the SLD itself but you also need to transfer  the PNGs on the target server.

File: src/core/qgsmaplayer.h (hunk 4 above)
Comment (elpaso) [reply src/core/qgsmaplayer.h]: done

**Comment by nyalldawson** [comment]
> Note that the my (possibly wrong) rationale behind the choice to not make tiles export by default is twofold: first I would like to keep the old behavior the default (because I guess that GeoServer is the most common target), second having the PNG as sidecar files makes it more difficult to use the SLD on the target because you don't just copy-paste the SLD itself but you also need to transfer the PNGs on the target server.

Ok, that's fair enough :+1:

**Pull request merged**